File: ios/RecipeClipper/UI/Recipes/RecipesViewModel.swift
import Combine
import Foundation
import Observation

/// `recipes` is nil until the database has answered, to tell "loading" from "nothing
/// matched". `pendingDeletes` holds the titles swiped away but not yet settled, newest last,
/// for the undo snackbar.
struct RecipesUiState: Equatable {
    var query = ""
    var recipes: [RecipeSummary]?
    var pendingDeletes: [String] = []
}

@MainActor
@Observable
final class RecipesViewModel {
    static let debounce = Duration.milliseconds(250)

    private(set) var uiState = RecipesUiState()

    @ObservationIgnored private let repository: RecipeRepository
    @ObservationIgnored private let sleep: Sleep
    @ObservationIgnored private var debounceTask: Task<Void, Never>?
    @ObservationIgnored private var appliedQuery: String?
    @ObservationIgnored private var historySubscription: AnyCancellable?

    // What swipe-to-delete captured, keyed by recipe id and kept in swipe order, so a second
    // swipe within the snackbar's few seconds cannot overwrite the first and strand it.
    @ObservationIgnored private var captured: [Int64: DeletedRecipe] = [:]
    @ObservationIgnored private var capturedOrder: [Int64] = []

    init(repository: RecipeRepository, sleep: @escaping Sleep = Sleeps.real) {
        self.repository = repository
        self.sleep = sleep
        schedule(query: "")
    }

    func onQueryChange(_ text: String) {
        uiState.query = text
        schedule(query: text)
    }

    /// Debounce, then drop a repeat of the query already applied, then switch to the latest
    /// query's results (the old subscription is cancelled) — Android's
    /// `debounce(250).distinctUntilChanged().flatMapLatest`.
    private func schedule(query: String) {
        debounceTask?.cancel()
        debounceTask = Task { [weak self, sleep] in
            do { try await sleep(Self.debounce) } catch { return }
            guard !Task.isCancelled, let self, query != self.appliedQuery else { return }
            self.appliedQuery = query
            self.historySubscription = self.repository.observeHistory(query: query)
                .receive(on: DispatchQueue.main)
                .sink { [weak self] recipes in self?.uiState.recipes = recipes }
        }
    }

    /// Swipe-to-delete: deletes immediately, keeping what it removed so `onUndoDelete` can
    /// restore it.
    func onDelete(_ recipe: RecipeSummary) {
        Task {
            guard let removed = await repository.delete(id: recipe.id) else { return }
            if captured[recipe.id] == nil { capturedOrder.append(recipe.id) }
            captured[recipe.id] = removed
            uiState.pendingDeletes.append(recipe.title)
        }
    }

    /// Restores everything still pending, oldest first. All or nothing: one snackbar covers
    /// the batch, so there is no way to pick one out of it.
    func onUndoDelete() {
        guard !capturedOrder.isEmpty else { return }
        let removed = capturedOrder.compactMap { captured[$0] }
        clearCaptured()
        Task {
            for deleted in removed { await repository.restore(deleted) }
        }
    }

    /// The snackbar timed out: the deletes stand.
    func onSnackbarDismissed() {
        clearCaptured()
    }

    private func clearCaptured() {
        captured.removeAll()
        capturedOrder.removeAll()
        uiState.pendingDeletes = []
    }
}

/// How long the undo snackbar stays, and what settles the batch when it goes. Run from the
/// screen's `.task(id: pendingDeletes)`, so it is restarted by every new delete and cancelled
/// when the screen goes away; Android's `LaunchedEffect(pendingDeletes)` around
/// `showSnackbar`.
///
/// On cancellation neither outcome runs — no undo, no dismissal — so a restart for a second
/// swipe leaves the first capture for the new snackbar's Undo, and leaving the screen neither
/// restores anything nor forgets it (coming back re-shows the snackbar, as on Android; popping
/// History drops the ViewModel and with it the captures, so the deletes stand).
enum SnackbarTimeout {
    static let duration = Duration.seconds(4)

    @MainActor
    static func run(pending: [String], sleep: Sleep = Sleeps.real, onTimeout: () -> Void) async {
        guard !pending.isEmpty else { return }
        do { try await sleep(duration) } catch { return }
        guard !Task.isCancelled else { return }
        onTimeout()
    }
}

File: app/src/main/java/com/example/recipeclipper/ui/recipes/RecipesViewModel.kt
package com.example.recipeclipper.ui.recipes

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.recipeclipper.data.RecipeRepository
import com.example.recipeclipper.data.model.RecipeSummary
import com.example.recipeclipper.ui.home.UrlInput
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import com.example.recipeclipper.data.LibraryPolicy
import com.example.recipeclipper.data.model.LibraryLimit
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.text.Collator
import javax.inject.Inject

/** How the Recipes screen orders its rows (#102). Held in memory, like the pantry's sort. */
enum class RecipeSort { RECENTLY_VIEWED, NAME, DATE_ADDED }

/**
 * [recipes] is null until the database has answered, to tell "loading" from "nothing
 * matched". [pendingDeletes] holds the titles swiped away but not yet settled, newest last,
 * for the undo snackbar; it survives rotation because it lives here, not in `remember`.
 * [linkInput] is the "Paste a link" dialog's text, or null while the dialog is closed.
 */
data class RecipesUiState(
    val query: String = "",
    val recipes: List<RecipeSummary>? = null,
    val pendingDeletes: List<String> = emptyList(),
    val sort: RecipeSort = RecipeSort.RECENTLY_VIEWED,
    val linkInput: String? = null,
    /** The free library's quiet count (#107); null when there is no limit to count against. */
    val count: LibraryCount? = null
) {
    /** The dialog's Open is enabled only for what Home's link field would open. */
    val canOpenLink: Boolean get() = linkInput?.let(UrlInput::normalize) != null
}

/** [recipes] saved of the free tier's [max]; a library from before the limit can be over it. */
data class LibraryCount(val recipes: Int, val max: Int) {
    val over: Boolean get() = recipes > max
}

@OptIn(FlowPreview::class, ExperimentalCoroutinesApi::class)
@HiltViewModel
class RecipesViewModel @Inject constructor(
    private val repository: RecipeRepository,
    library: LibraryPolicy = LibraryPolicy.HistoryOnly
) : ViewModel() {

    // The query box's live value. Kept in a StateFlow, not `remember`, so it survives rotation.
    private val query = MutableStateFlow("")
    private val pendingDeletes = MutableStateFlow<List<String>>(emptyList())
    private val sort = MutableStateFlow(RecipeSort.RECENTLY_VIEWED)
    private val linkInput = MutableStateFlow<String?>(null)

    // What swipe-to-dismiss captured, keyed by recipe id so a second swipe within the
    // snackbar's few seconds cannot overwrite the first and strand it. A plain field: the
    // ViewModel already survives rotation, the same reason RecipeViewModel keeps its timer
    // deadlines this way.
    private val captured = LinkedHashMap<Long, RecipeRepository.DeletedRecipe>()

    private val recipes: StateFlow<List<RecipeSummary>?> = query
        .debounce(250)
        .distinctUntilChanged()
        .flatMapLatest { repository.observeHistory(it) }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), null)

    // Only on the free tier: unlocked, or with the flag off, there is nothing to count against.
    private val count = combine(repository.observeCount(), library.limits) { saved, limit ->
        (limit as? LibraryLimit.Free)?.let { LibraryCount(saved, it.max) }
    }

    val uiState: StateFlow<RecipesUiState> =
        combine(
            combine(query, recipes, pendingDeletes, sort, linkInput) { q, found, pending, order, link ->
                RecipesUiState(q, found?.let { sorted(it, order) }, pending, order, link)
            },
            count
        ) { state, saved -> state.copy(count = saved) }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), RecipesUiState())

    fun onQueryChange(text: String) {
        query.value = text
    }

    fun onSortChange(value: RecipeSort) {
        sort.value = value
    }

    /** "Paste a link" from the + menu: opens the dialog, empty. */
    fun onPasteLink() {
        linkInput.value = ""
    }

    fun onLinkChange(text: String) {
        linkInput.value = text
    }

    fun onLinkDismiss() {
        linkInput.value = null
    }

    /** The link to import, closing the dialog; null, and the dialog stays, if it isn't one. */
    fun onOpenLink(): String? {
        val url = linkInput.value?.let(UrlInput::normalize) ?: return null
        linkInput.value = null
        return url
    }

    /** Swipe-to-dismiss: deletes immediately, keeping what it removed so [onUndoDelete] can restore it. */
    fun onDelete(recipe: RecipeSummary) {
        viewModelScope.launch {
            val removed = repository.delete(recipe.id) ?: return@launch
            captured[recipe.id] = removed
            pendingDeletes.value = pendingDeletes.value + recipe.title
        }
    }

    /**
     * Restores everything still pending, oldest first so ids go back in the order they left.
     * All or nothing: one snackbar covers the batch, so there is no way to pick one out of it.
     */
    fun onUndoDelete() {
        if (captured.isEmpty()) return
        val removed = captured.values.toList()
        captured.clear()
        pendingDeletes.value = emptyList()
        viewModelScope.launch { removed.forEach { repository.restore(it) } }
    }

    /** The snackbar timed out or was swiped away: the deletes stand. */
    fun onSnackbarDismissed() {
        captured.clear()
        pendingDeletes.value = emptyList()
    }

    companion object {
        /**
         * [recipes], which the database gives newest viewed first, in [sort]'s order. Name
         * follows the phone's language (a collator, not code points); Date added is newest
         * first by id, which only ever grows (AUTOINCREMENT). Stable, so ties keep recency.
         */
        fun sorted(recipes: List<RecipeSummary>, sort: RecipeSort): List<RecipeSummary> = when (sort) {
            RecipeSort.RECENTLY_VIEWED -> recipes
            RecipeSort.NAME -> recipes.sortedWith(compareBy(Collator.getInstance()) { it.title })
            RecipeSort.DATE_ADDED -> recipes.sortedByDescending { it.id }
        }
    }
}
